src/Orleans/Statistics/CounterStatistic.cs

```diff
@@ -179,11 +185,27 @@ public void IncrementBy(long n)
         {
             if (isOrleansManagedThread)
             {
-                while (perOrleansThreadCounters.Count <= id)
+                ReferenceLong orleansValue = null;
+               
+                try
```

Comment:
One CPU branch will cost nothing, compared to the code complexity.
An even more radical alternative is to use Dictionary per thread with a key being counter name. That will remove the need to re-grow the list. But that would introduce dictionary lookup on every increment instead of List access, which is more expensive. 
But one if is nothing.